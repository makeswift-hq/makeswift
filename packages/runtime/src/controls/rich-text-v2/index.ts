export * from './rich-text-v2'
export * from './translation'
