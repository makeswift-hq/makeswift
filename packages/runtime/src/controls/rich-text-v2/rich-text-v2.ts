import { PropController } from '../../prop-controllers/base'
import { Descendant, Editor, Transforms } from 'slate'
import { ReactEditor, RenderElementProps, RenderLeafProps } from 'slate-react'
import { BoxModel } from '../../box-model'
import { ReactNode, KeyboardEvent } from 'react'
import { Send } from '../../prop-controllers/instances'

export type RichTextV2ControlData = Descendant[]

export const RichTextV2ControlType = 'makeswift::controls::rich-text-v2'

export const RichTextV2Mode = {
  Inline: 'makeswift::controls::rich-text-v2::mode::inline',
  Block: 'makeswift::controls::rich-text-v2::mode::block',
} as const

export type RichTextV2Mode = typeof RichTextV2Mode[keyof typeof RichTextV2Mode]

export type RichTextV2Plugin = {
  withPlugin?(editor: Editor): Editor
  onKeyDown?(event: KeyboardEvent): void
  renderLeaf?(props: RenderLeafProps): ReactNode
  renderElement?(props: RenderElementProps): ReactNode
}

type RichTextV2Config = {
  plugins?: RichTextV2Plugin[]
  mode?: RichTextV2Mode
}

export type RichTextV2ControlDefinition<T extends RichTextV2Config = RichTextV2Config> = {
  type: typeof RichTextV2ControlType
  config: T
}

export function unstable_RichTextV2<T extends RichTextV2Config>(
  config: T = {} as T,
): RichTextV2ControlDefinition {
  return { type: RichTextV2ControlType, config }
}

export const RichTextV2ControlMessageType = {
  // COSMOS -> HOST
  RESET_VALUE: 'RESET_VALUE',
  FOCUS: 'FOCUS',

  // HOST -> COSMOS
  SET_DEFAULT_VALUE: 'SET_DEFAULT_VALUE',
  ON_CHANGE: 'ON_CHANGE',
  SELECT: 'SELECT',
  SWITCH_TO_BUILD_MODE: 'SWITCH_TO_BUILD_MODE',
  CHANGE_BOX_MODEL: 'CHANGE_BOX_MODEL',
} as const

type OnChangeRichTextControlMessage = {
  type: typeof RichTextV2ControlMessageType.ON_CHANGE
  value: RichTextV2ControlData
}

type SetDefaultValueRichTextControlMessage = {
  type: typeof RichTextV2ControlMessageType.SET_DEFAULT_VALUE
  value: RichTextV2ControlData
}

type ResetValueRichTextControlMessage = { type: typeof RichTextV2ControlMessageType.RESET_VALUE }

type SelectRichTextControlMessage = { type: typeof RichTextV2ControlMessageType.SELECT }

type FocusRichTextControlMessage = { type: typeof RichTextV2ControlMessageType.FOCUS }

type SwitchToBuildModeRichTextControlMessage = {
  type: typeof RichTextV2ControlMessageType.SWITCH_TO_BUILD_MODE
}

type BoxModelChangeRichControlMessage = {
  type: typeof RichTextV2ControlMessageType.CHANGE_BOX_MODEL
  payload: { boxModel: BoxModel | null }
}

export type RichTextV2ControlMessage =
  | OnChangeRichTextControlMessage
  | SetDefaultValueRichTextControlMessage
  | ResetValueRichTextControlMessage
  | FocusRichTextControlMessage
  | SelectRichTextControlMessage
  | SwitchToBuildModeRichTextControlMessage
  | BoxModelChangeRichControlMessage

export class RichTextV2Control<
  T extends RichTextV2ControlDefinition = RichTextV2ControlDefinition,
> extends PropController<RichTextV2ControlMessage> {
  private editor: Editor | null = null
  private defaultValue: RichTextV2ControlData | null = null
  descriptor: RichTextV2ControlDefinition

  constructor(send: Send<RichTextV2ControlMessage>, descriptor: T) {
    super(send)

    this.descriptor = descriptor
    this.send = send
  }

  recv = (message: RichTextV2ControlMessage): void => {
    if (!this.editor) return
    switch (message.type) {
      case RichTextV2ControlMessageType.FOCUS: {
        ReactEditor.focus(this.editor)
        Transforms.select(this.editor, {
          anchor: Editor.start(this.editor, []),
          focus: Editor.end(this.editor, []),
        })
        break
      }
      case RichTextV2ControlMessageType.RESET_VALUE: {
        if (this.defaultValue) {
          this.editor.selection = null
          this.editor.children = this.defaultValue
          this.editor.onChange()
        }
      }
    }
  }

  setEditor(editor: Editor) {
    this.editor = editor
  }

  setDefaultValue(defaultValue: RichTextV2ControlData) {
    this.defaultValue = defaultValue
    this.send({ type: RichTextV2ControlMessageType.SET_DEFAULT_VALUE, value: defaultValue })
  }

  select() {
    this.send({ type: RichTextV2ControlMessageType.SELECT })
  }

  switchToBuildMode() {
    this.send({ type: RichTextV2ControlMessageType.SWITCH_TO_BUILD_MODE })
  }

  onChange(value: Descendant[]) {
    this.send({
      type: RichTextV2ControlMessageType.ON_CHANGE,
      value,
    })
  }

  changeBoxModel(boxModel: BoxModel | null): void {
    this.send({ type: RichTextV2ControlMessageType.CHANGE_BOX_MODEL, payload: { boxModel } })
  }
}
