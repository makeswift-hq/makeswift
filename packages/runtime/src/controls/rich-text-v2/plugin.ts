import { type KeyboardEvent } from 'react'
import { type Editor } from 'slate'
import { type RenderElementProps, type RenderLeafProps } from 'slate-react'
import { ControlDefinition, type Data, type RichTextPluginControl } from '@makeswift/controls'

export type RenderElement = (props: RenderElementProps) => JSX.Element
export type RenderLeaf = (props: RenderLeafProps) => JSX.Element

export type RichTextV2Plugin<C extends RichTextPluginControl = RichTextPluginControl> = {
  control?: C
  withPlugin?(editor: Editor): Editor
  onKeyDown?(event: KeyboardEvent, editor: Editor): void
  renderElement?: (
    renderElement: RenderElement,
    value: any,
  ) => (props: RenderElementProps) => JSX.Element
  renderLeaf?: (renderLeaf: RenderLeaf, value: any) => (props: RenderLeafProps) => JSX.Element
}

export function Plugin<Def extends ControlDefinition, ValueType extends Data>({
  control,
  withPlugin,
  onKeyDown,
  renderElement,
  renderLeaf,
}: RichTextV2Plugin<RichTextPluginControl<Def, ValueType>>) {
  return { control, withPlugin, onKeyDown, renderElement, renderLeaf }
}
