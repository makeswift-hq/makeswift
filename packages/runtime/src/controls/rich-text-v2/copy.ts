import { Descendant, Element, Text } from 'slate'
import { Inline } from '../../slate'
import { ElementUtils } from '../../slate/utils/element'
import { CopyContext } from '../../state/react-page'
import { RichTextV2ControlData } from './rich-text-v2'
import { richTextV2DescendentsToData } from './dto'

export function copyRichTextV2Data(
  data: RichTextV2ControlData | undefined,
  context: CopyContext,
): RichTextV2ControlData | undefined {
  if (data == null) return data

  return richTextV2DescendentsToData(data.descendants?.flatMap(d => copyNode(d)))

  function copyNode(descendant: Descendant): Descendant {
    if (Element.isElement(descendant) && ElementUtils.isInline(descendant)) {
      return copyLinkData(descendant)
    }

    if (Text.isText(descendant)) {
      return copyTypographyData(descendant)
    }

    return { ...descendant, children: descendant.children.map(copyNode) }
  }

  function copyLinkData(descendant: Inline): Inline {
    if (descendant.type !== 'link') {
      return { ...descendant, children: descendant.children.map(copyNode) as (Inline | Text)[] }
    }

    if (descendant.link == null) return descendant

    switch (descendant.link.type) {
      case 'OPEN_PAGE': {
        const pageId = descendant.link.payload.pageId

        if (pageId == null) return descendant

        return {
          ...descendant,
          link: {
            ...descendant.link,
            payload: {
              ...descendant.link.payload,
              pageId:
                context.replacementContext.pageIds.get(pageId) ?? descendant.link.payload.pageId,
            },
          },
          children: descendant.children.map(copyNode) as (Inline | Text)[],
        }
      }

      case 'SCROLL_TO_ELEMENT': {
        const elementIdConfig = descendant.link.payload.elementIdConfig

        if (elementIdConfig == null) return descendant

        return {
          ...descendant,
          link: {
            ...descendant.link,
            payload: {
              ...descendant.link.payload,
              elementIdConfig: {
                ...elementIdConfig,
                elementKey:
                  context.replacementContext.elementKeys.get(elementIdConfig.elementKey) ??
                  elementIdConfig.elementKey,
              },
            },
          },
          children: descendant.children.map(copyNode) as (Inline | Text)[],
        }
      }

      default:
        return descendant
    }
  }

  function copyTypographyData(descendant: Text): Text {
    return {
      ...descendant,

      ...(descendant.typography == null
        ? {}
        : {
            typography: {
              id:
                context.replacementContext.typographyIds.get(descendant.typography?.id ?? '') ??
                descendant.typography?.id,
              style: descendant.typography?.style.map(override => ({
                ...override,
                value: {
                  ...override.value,
                  ...(override.value.color == null
                    ? {}
                    : {
                        color: {
                          ...override.value.color,
                          swatchId:
                            context.replacementContext.swatchIds.get(
                              override.value.color?.swatchId ?? '',
                            ) ?? override.value.color?.swatchId,
                        },
                      }),
                },
              })),
            },
          }),
    }
  }
}
