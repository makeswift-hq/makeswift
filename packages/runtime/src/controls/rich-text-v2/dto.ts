import { Descendant } from 'slate'
import { RichTextV2ControlData, RichTextV2ControlType } from './rich-text-v2'
import { v4 as uuid } from 'uuid'
import { BlockType } from '../../slate/types'

export function richTextV2DataToDescendents(data: RichTextV2ControlData): Descendant[] {
  return data.descendants.length === 0
    ? [{ type: BlockType.Default, children: [{ text: '' }] }]
    : data.descendants
}

export function richTextV2DescendentsToData(
  descendants: Descendant[],
  key?: string,
): RichTextV2ControlData {
  return {
    type: RichTextV2ControlType,
    version: 2,
    descendants,
    key: key ?? uuid(),
  }
}
