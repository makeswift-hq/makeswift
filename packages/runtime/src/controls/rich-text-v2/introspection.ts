import { Data } from '@makeswift/controls'
import { getPageIds, getSwatchIds, getTypographyIds } from '../../prop-controllers/introspection'
import { ControlDefinition } from '../control'
import { RichTextV2ControlData, RichTextV2ControlDefinition } from './rich-text-v2'
import { Descendant, Element as SlateElement, Text } from 'slate'

function introspectRichTextV2Data<T>(
  definition: RichTextV2ControlDefinition,
  data: RichTextV2ControlData,
  func: (definition: ControlDefinition, data: Data) => T[],
): T[] {
  const plugins = definition.config.plugins

  return data.descendants.flatMap(d => getDescendantTypographyIds(d))

  function getDescendantTypographyIds(descendant: Descendant): T[] {
    if (SlateElement.isElement(descendant)) {
      return [
        ...getElementTypographyIds(descendant),
        ...descendant.children.flatMap(d => getDescendantTypographyIds(d)),
      ]
    }

    if (Text.isText(descendant)) {
      return getTextTypographyIds(descendant)
    }

    return []
  }

  function getElementTypographyIds(descendant: SlateElement) {
    return (
      plugins?.flatMap(plugin =>
        plugin.control?.definition && plugin.control.getElementValue
          ? func(plugin.control.definition, plugin.control.getElementValue(descendant))
          : [],
      ) ?? []
    )
  }

  function getTextTypographyIds(text: Text) {
    return (
      plugins?.flatMap(plugin => {
        return plugin.control?.definition && plugin.control.getLeafValue
          ? func(plugin.control.definition, plugin.control.getLeafValue(text))
          : []
      }) ?? []
    )
  }
}

export function getRichTextV2SwatchIds(
  definition: RichTextV2ControlDefinition,
  data: RichTextV2ControlData,
) {
  return introspectRichTextV2Data(definition, data, getSwatchIds)
}

export function getRichTextV2TypographyIds(
  definition: RichTextV2ControlDefinition,
  data: RichTextV2ControlData,
) {
  return introspectRichTextV2Data(definition, data, getTypographyIds)
}

export function getRichTextV2PageIds(
  definition: RichTextV2ControlDefinition,
  data: RichTextV2ControlData,
) {
  return introspectRichTextV2Data(definition, data, getPageIds)
}
