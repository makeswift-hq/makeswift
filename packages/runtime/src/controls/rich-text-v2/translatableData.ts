import { Descendant, Text } from 'slate'
import escapeHtml from 'escape-html'
import { RichTextV2ControlData, RichTextV2ControlDefinition } from './rich-text-v2'
import { InlineType } from '../../slate/types'

function hashTranslationString(s: string): string {
  let hash = 0
  for (let i = 0, len = s.length; i < len; i++) {
    let chr = s.charCodeAt(i)
    hash = (hash << 5) - hash + chr
    hash |= 0 // Convert to 32bit integer
  }
  return hash.toString()
}

function getRichTextV2TranslatableDescendant(
  definition: RichTextV2ControlDefinition,
  descendant: Descendant,
): string {
  if (Text.isText(descendant)) {
    let string = escapeHtml(descendant.text)
    return string
  }

  const children = descendant.children
    .map((child: Descendant) => getRichTextV2TranslatableDescendant(definition, child))
    .join('')

  switch (descendant.type) {
    case InlineType.Link:
      return `<a>${children}</a>`
    case InlineType.Link:
      return `<sup>${children}</sup>`
    case InlineType.Link:
      return `<sub>${children}</sub>`
    case InlineType.Link:
      return `<code>${children}</code>`
    default:
      return children
  }
}

export function getRichTextV2TranslatableData(
  definition: RichTextV2ControlDefinition,
  data: RichTextV2ControlData,
): Record<string, string> {
  return data.descendants.reduce((acc, descendant: Descendant) => {
    const value = getRichTextV2TranslatableDescendant(definition, descendant)

    const key = hashTranslationString(value)

    return { ...acc, [key]: value }
  }, {})
}
