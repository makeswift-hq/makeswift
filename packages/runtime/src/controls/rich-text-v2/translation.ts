import { Descendant } from 'slate'
import { RichTextV2ControlData, RichTextV2ControlType } from './rich-text-v2'
import { v4 as uuid } from 'uuid'

export function richTextV2DataToDescendents(data: RichTextV2ControlData): Descendant[] {
  return data.descendants
}

export function richTextV2DescendentsToData(
  descendants: Descendant[],
  key?: string,
): RichTextV2ControlData {
  return {
    type: RichTextV2ControlType,
    version: 2,
    descendants,
    key: key ?? uuid(),
  }
}
