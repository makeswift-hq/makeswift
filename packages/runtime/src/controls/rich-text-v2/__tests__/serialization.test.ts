import { deserializeRecord } from '@makeswift/controls'
import { RichText, RichTextV2Definition } from '../rich-text-v2'
import { deserializeUnifiedControlDef } from '../../../builder/serialization/control-serialization'

jest.mock('corporate-ipsum', () => {
  return {
    __esModule: true,
    default: jest.fn(() => '[generated random text]'),
  }
})

expect.addSnapshotSerializer({
  serialize: () => '[MessagePort]',
  test: value => value instanceof MessagePort,
})

describe('RichTextV2 serialization', () => {
  test.each([
    ['block', RichText()],
    ['inline', RichText({ mode: RichText.Mode.Inline })],
  ])('deserialized definition in %s mode contains correct plugins', (_mode, def) => {
    const [serializedDefinition, transferables] = def.serialize()
    expect(serializedDefinition).toMatchSnapshot('serialized definition')
    expect(transferables.length).toMatchSnapshot('transferables')
    transferables.forEach((port: any) => port.close())

    RichTextV2Definition.generateParagraph = jest.fn(() => 'generated paragraph')
    const deserializedDefinition = RichTextV2Definition.deserialize(
      deserializeRecord(serializedDefinition),
      deserializeUnifiedControlDef,
    )

    expect(deserializedDefinition).toMatchSnapshot()
  })
})
