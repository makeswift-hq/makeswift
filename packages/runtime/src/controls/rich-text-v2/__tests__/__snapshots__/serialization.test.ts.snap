// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RichTextV2 serialization deserialized definition in block mode contains correct plugins 1`] = `
RichTextV2Definition {
  "config": {
    "defaultValue": "[generated random text]",
    "mode": undefined,
    "plugins": [
      {
        "control": {
          "definition": SelectDefinition {
            "config": {
              "defaultValue": "paragraph",
              "label": "Block",
              "labelOrientation": "horizontal",
              "options": [
                {
                  "label": "Paragraph",
                  "value": "paragraph",
                },
                {
                  "label": "Heading 1",
                  "value": "heading-one",
                },
                {
                  "label": "Heading 2",
                  "value": "heading-two",
                },
                {
                  "label": "Heading 3",
                  "value": "heading-three",
                },
                {
                  "label": "Heading 4",
                  "value": "heading-four",
                },
                {
                  "label": "Heading 5",
                  "value": "heading-five",
                },
                {
                  "label": "Heading 6",
                  "value": "heading-six",
                },
                {
                  "label": "Bulleted list",
                  "value": "unordered-list",
                },
                {
                  "label": "Numbered list",
                  "value": "ordered-list",
                },
                {
                  "label": "Quote",
                  "value": "blockquote",
                },
              ],
            },
          },
        },
      },
      {
        "control": {
          "definition": Typography {
            "config": {},
          },
        },
      },
      {
        "control": {
          "definition": StyleV2Definition {
            "config": {
              "getStyle": [Function],
              "type": IconRadioGroupDefinition {
                "config": {
                  "defaultValue": "left",
                  "label": "Alignment",
                  "options": [
                    {
                      "icon": "TextAlignLeft",
                      "label": "Left Align",
                      "value": "left",
                    },
                    {
                      "icon": "TextAlignCenter",
                      "label": "Center Align",
                      "value": "center",
                    },
                    {
                      "icon": "TextAlignRight",
                      "label": "Right Align",
                      "value": "right",
                    },
                    {
                      "icon": "TextAlignJustify",
                      "label": "Justify",
                      "value": "justify",
                    },
                  ],
                },
              },
            },
          },
        },
      },
      {
        "control": {
          "definition": IconRadioGroupDefinition {
            "config": {
              "label": "Inline",
              "options": [
                {
                  "icon": "Superscript16",
                  "label": "Superscript",
                  "value": "superscript",
                },
                {
                  "icon": "Subscript16",
                  "label": "Subscript",
                  "value": "subscript",
                },
                {
                  "icon": "Code16",
                  "label": "Code",
                  "value": "code",
                },
              ],
            },
          },
        },
      },
      {
        "control": {
          "definition": LinkDefinition {
            "config": {
              "label": "On Click",
            },
          },
        },
      },
    ],
  },
}
`;

exports[`RichTextV2 serialization deserialized definition in block mode contains correct plugins: serialized definition 1`] = `
{
  "config": {
    "defaultValue": "[generated random text]",
    "mode": undefined,
    "plugins": [
      {
        "control": {
          "definition": {
            "config": {
              "defaultValue": "paragraph",
              "label": "Block",
              "labelOrientation": "horizontal",
              "options": [
                {
                  "label": "Paragraph",
                  "value": "paragraph",
                },
                {
                  "label": "Heading 1",
                  "value": "heading-one",
                },
                {
                  "label": "Heading 2",
                  "value": "heading-two",
                },
                {
                  "label": "Heading 3",
                  "value": "heading-three",
                },
                {
                  "label": "Heading 4",
                  "value": "heading-four",
                },
                {
                  "label": "Heading 5",
                  "value": "heading-five",
                },
                {
                  "label": "Heading 6",
                  "value": "heading-six",
                },
                {
                  "label": "Bulleted list",
                  "value": "unordered-list",
                },
                {
                  "label": "Numbered list",
                  "value": "ordered-list",
                },
                {
                  "label": "Quote",
                  "value": "blockquote",
                },
              ],
            },
            "type": "makeswift::controls::select",
          },
          "getValue": [MessagePort],
          "onChange": [MessagePort],
        },
      },
      {
        "control": {
          "definition": {
            "config": {},
            "type": "makeswift::controls::typography",
          },
          "getValue": [MessagePort],
          "onChange": [MessagePort],
        },
      },
      {
        "control": {
          "definition": {
            "config": {
              "getStyle": [MessagePort],
              "type": {
                "config": {
                  "defaultValue": "left",
                  "label": "Alignment",
                  "options": [
                    {
                      "icon": "TextAlignLeft",
                      "label": "Left Align",
                      "value": "left",
                    },
                    {
                      "icon": "TextAlignCenter",
                      "label": "Center Align",
                      "value": "center",
                    },
                    {
                      "icon": "TextAlignRight",
                      "label": "Right Align",
                      "value": "right",
                    },
                    {
                      "icon": "TextAlignJustify",
                      "label": "Justify",
                      "value": "justify",
                    },
                  ],
                },
                "type": "makeswift::controls::icon-radio-group",
              },
            },
            "type": "makeswift::controls::style-v2",
          },
          "getValue": [MessagePort],
          "onChange": [MessagePort],
        },
      },
      {
        "control": {
          "definition": {
            "config": {
              "label": "Inline",
              "options": [
                {
                  "icon": "Superscript16",
                  "label": "Superscript",
                  "value": "superscript",
                },
                {
                  "icon": "Subscript16",
                  "label": "Subscript",
                  "value": "subscript",
                },
                {
                  "icon": "Code16",
                  "label": "Code",
                  "value": "code",
                },
              ],
            },
            "type": "makeswift::controls::icon-radio-group",
          },
          "getValue": [MessagePort],
          "onChange": [MessagePort],
        },
      },
      {
        "control": {
          "definition": {
            "config": {
              "label": "On Click",
            },
            "type": "makeswift::controls::link",
          },
          "getValue": [MessagePort],
          "onChange": [MessagePort],
        },
      },
    ],
  },
  "type": "makeswift::controls::rich-text-v2",
}
`;

exports[`RichTextV2 serialization deserialized definition in block mode contains correct plugins: transferables 1`] = `11`;

exports[`RichTextV2 serialization deserialized definition in inline mode contains correct plugins 1`] = `
RichTextV2Definition {
  "config": {
    "defaultValue": "Edit this text",
    "mode": "makeswift::controls::rich-text-v2::mode::inline",
    "plugins": [
      {},
    ],
  },
}
`;

exports[`RichTextV2 serialization deserialized definition in inline mode contains correct plugins: serialized definition 1`] = `
{
  "config": {
    "defaultValue": "Edit this text",
    "mode": "makeswift::controls::rich-text-v2::mode::inline",
    "plugins": [
      {},
    ],
  },
  "type": "makeswift::controls::rich-text-v2",
}
`;

exports[`RichTextV2 serialization deserialized definition in inline mode contains correct plugins: transferables 1`] = `0`;
